src/Image.hs

```diff
@@ -30,10 +36,38 @@ histogram arr = hist
                          let r = levelAt pos 0
                              g = levelAt pos 1
                              b = levelAt pos 2
-                         in (incElem r g b hist))
+                         in incElem r g b hist)
                          zero
                          imgMatrix
-           
+
+histogram' :: M.RawImage -> M.Palette
+histogram' pixels = topColors
+  where
+    buckets = createBuckets pixels histogramGridSize
+    avgColorsAndCounts = map (\bucket -> (C.avarageColor bucket, length bucket)) buckets
+    sorted = L.sortBy (\(_, count1) (_, count2) -> compare count2 count1) avgColorsAndCounts
+    topColors = take paletteSize $ map (C.rgbToHex . fst) sorted
+
+createBuckets :: M.RawImage -> Int -> [[C.RGB]]
+createBuckets pixels gridSize = Map.elems $ execState createBuckets' emptyState
+  where
+    (Z :. nrRow :. nrCol :. _) = R.extent pixels
+    imgMatrix = [(row, col) | row <- [0..nrRow - 1], col <- [0..nrCol - 1]]
+    emptyState = Map.empty
+
+    levelAt :: (Int, Int) -> Int -> Int
+    levelAt (row, col) rgb = fromIntegral $ R.unsafeIndex pixels (Z :. row :. col :. rgb)
+
+    createBuckets' :: State (Map.Map Int [C.RGB]) ()
```

Comment:
just `buckets` - create to ta lambda chyba z pos rozumiem

Reply:
Tego kom nie rozumiem

Reply:
to nie jest cos co robi createbuckets, tylko to jest juz policzony state z buckets

Reply:
czaję, ale nie mam teraz weny na nową nazwę